nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessor.java

```diff
@@ -239,63 +246,151 @@ private WorkflowActionType getActionType(StateExecutionImpl execution) {
   }
 
   private boolean isNextActivationImmediately(StateExecutionImpl execution) {
-    return execution.getNextActivation() != null && !execution.getNextActivation().isAfterNow();
+    return execution.isStateProcessInvoked() && execution.getNextActivation() != null && !execution.getNextActivation().isAfterNow();
   }
 
-  private NextAction processState(WorkflowInstance instance, AbstractWorkflowDefinition<?> definition,
-      StateExecutionImpl execution, WorkflowState currentState) {
-    WorkflowStateMethod method = definition.getMethod(instance.state);
-    if (method == null) {
-      execution.setNextState(currentState);
-      return stopInState(currentState, "Execution finished.");
+  private NextAction processWithListeners(ListenerContext listenerContext, WorkflowInstance instance,
+      AbstractWorkflowDefinition<? extends WorkflowState> definition, StateExecutionImpl execution, WorkflowState state) {
+    List<WorkflowExecutorListener> chain = new LinkedList<>(asList(this.executorListeners));
```

Comment:
Prefer arraylist (with known size + 1 for the bonus element) to linked list.
The type of this.executorListeners could already be an Collection. No need to convert it on every invocation.